verifier/verifier.go

```diff
@@ -46,99 +46,216 @@ import (
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-// verifier implements notation.Verifier and notation.verifySkipper
-type verifier struct {
-	trustPolicyDoc   *trustpolicy.Document
-	trustStore       truststore.X509TrustStore
-	pluginManager    plugin.Manager
-	revocationClient revocation.Revocation
+// Verifier implements notation.Verifier, notation.BlobVerifier and notation.verifySkipper
+type Verifier struct {
```

Comment:
To avoid creating duplicate verifiers for OCI and blob verification purposes, we can leverage the fact that the underlying struct already implements both notation.BlobVerifier and notation.Verifier interfaces.

Reply:
Not really, we aren't exporting any field inside the struct so user cannot modify any field.

Reply:
I see. Then keeping it as `unexported` should be fine? Is there any blocker that requires the struct must be exported? 

I'm thinking if we should provide clear constructors for OCI and blob verifiers respectively. By 'clear' I mean unmixed constructors, user should know exactly what they want and create. The `NewVerifier` constructor in this PR still looks a bit error prone to me (it's possible that user chooses to only pass in an OCI trust policy, but later decides to use it as a blob verifier interface, then their verification would have an issue). Maybe others have different opinions? /cc: @shizhMSFT @JeyJeyGao

Reply:
A public method that returns unexported/private struct seems odd.

We are providing both kind of constroctors i.e. oci only, blob only and oci&blob only, each has there own usecase. If user want oci only then they should use oci only constroctor.

Reply:
> A public method that returns unexported/private struct seems odd.

We are not returning the unexported/private struct `verifier` here.
For OCI, the constructor would be 
```
func NewOciVerifierWithOptions(ociTrustPolicy *trustpolicy.OCIDocument, trustStore truststore.X509TrustStore, pluginManager plugin.Manager, opts VerifierOptions) (notation.Verifier, error)   // user cannot pass a blob trust policy to an OCI verifier.

func NewOciVerifierFromConfig() (notation.Verifier, error)
```

For Blob, the constructor would be 
```
func NewBlobVerifierWithOptions(blobTrustPolicy *trustpolicy.BlobDocument, trustStore truststore.X509TrustStore, pluginManager plugin.Manager, opts VerifierOptions) (notation.BlobVerifier , error)    // user cannot pass an oci trust policy to a blob verifier.

func NewBlobVerifierFromConfig() (notation.BlobVerifier , error)
```

We shouldn't provide a constructor of `oci + blob` by returning the concrete struct. It's error prone, because semantically, due to Golang duck typing, it's possible that the caller only passed in an OCI trust policy but use it as a `notatoin.BlobVerifier ` later, this is a semantic error and Notation is not checking it. This kind of error is implicit and hard to figure out for the caller during runtime.

Reply:
I think a public method that returns an unexported struct as an interface is a common practice to hide the internal data structure.

Reply:
It is better to keep it as a private struct. We should export them as little as possible.